File: day18/day18.py
#!/usr/bin/env python3

import sys

uppercase_alpha = bytes(list(range(ord('A'),ord('Z')+1))).decode()
lowercase_alpha = bytes(list(range(ord('a'),ord('z')+1))).decode()

def get_surrounding_positions(pos, height, width):

    valid_pos = []
    for dim in [0,1]:
        for o in [-1,1]:
            newpos = [pos[0],pos[1]]
            newpos[dim] += o
            if newpos[0] >= 0 and newpos[0] < height and newpos[1] >= 0 and newpos[1] < width:
                valid_pos.append(newpos)
    return valid_pos


class Map:
    def __init__(self,arg):
        if type(arg) == Map:
            self.teh_map = [list(line) for line in arg.teh_map]
            self.height = arg.height
            self.width = arg.width
        else:
            teh_map = arg
            self.teh_map = [list(line) for line in teh_map]
            self.height = len(teh_map)
            self.width = len(teh_map[0])

    def print(self):
        for l in self.teh_map:
            print(''.join(l))

    def get(self,pt):
        return self.teh_map[pt[0]][pt[1]]

    def set(self,pt,t):
        self.teh_map[pt[0]][pt[1]] = t

    def get_tile_locations(self,t):
        return [[i,j] for i in range(self.height) for j in range(self.width) if self.get([i,j])==t]

    def get_accessible_keys(self,pos):

        mapp = Map(self)
        found_keys = []
        wavefronts = [pos]
        nsteps = 1

        while len(wavefronts) > 0:

            wavefronts_new = []
            for wpos in wavefronts:

                sur = get_surrounding_positions(wpos,self.height,self.width)
                for pt in sur:
                    t = mapp.get(pt)

                    if t == '#' or t in uppercase_alpha:
                        continue
                    if t in lowercase_alpha:
                        # we found a key!
                        found_keys.append([t,pt,nsteps])
                    if t == '.':
                        mapp.set(pt,'@')
                        # line = mapp[pt[0]]
                        # mapp[pt[0]] = line[0:pt[1]]+'@'+line[pt[1]+1:]
                        wavefronts_new.append(pt)

            wavefronts = wavefronts_new
            nsteps += 1

        return found_keys


def get_shortest_path_bruteforce(mapp, pos):

    accessible_keys = mapp.get_accessible_keys(pos)

    paths = []

    for ak in accessible_keys:

        key, kpos, kdist = ak
        mapp_copy = Map(mapp)

        # move position and remove key
        mapp_copy.set(pos,'.')
        mapp_copy.set(kpos,'@')
        # remove door/block
        door_locations = mapp_copy.get_tile_locations(key.upper())
        for dl in door_locations:
            mapp_copy.set(dl,'.')

        keys,dist = get_shortest_path_bruteforce(mapp_copy,kpos)
        paths.append([[key]+keys,dist+kdist])

    if len(paths) == 0:
        return [[],0]

    paths.sort(key=lambda p: p[1])

    return paths[0]



def compute_distances(labels, mapp):

    distances = {}
    for lbl in labels:
        startpos = mapp.get_tile_locations(lbl)[0]

        mapp_cpy = Map(mapp)
        mapp_cpy.set(startpos,'*')
        wavefronts = [(startpos,set())]
        nsteps = 1

        while len(wavefronts) > 0:

            wavefronts_new = []
            for wpos,doors_passed in wavefronts:

                sur = get_surrounding_positions(wpos,mapp_cpy.height,mapp_cpy.width)
                for pt in sur:
                    t = mapp_cpy.get(pt)

                    if t in lowercase_alpha:
                        # we found a key!
                        if t > lbl:
                            distances[lbl+t] = (nsteps,doors_passed)
                    if t != '*' and t!= '#':

                        if t in uppercase_alpha:
                            doors_passed = set(doors_passed)
                            doors_passed.add(t)

                        mapp_cpy.set(pt,'*')
                        # line = mapp[pt[0]]
                        # mapp[pt[0]] = line[0:pt[1]]+'@'+line[pt[1]+1:]
                        wavefronts_new.append((pt,doors_passed))

            wavefronts = wavefronts_new
            nsteps += 1


    return distances

def key_str(start,tgtset):
    return start+':'+''.join(sorted(tgtset))

def get_optimal_path(dst, tgtset, start, acquired_keys, distances):
    key = key_str(start,tgtset)

    if not key in dst:

        if len(tgtset) == 1:
            for tgt in tgtset:
                pair = ''.join(sorted([start,tgt]))
                distance,required_doors = distances[pair]
                required_keys = {d.lower() for d in required_doors}
                path = start+tgt
                if not required_keys.issubset(acquired_keys):
                    distance = None

        else:

            distance = None
            path = None

            for tgt in tgtset:

                pair = ''.join(sorted([start,tgt]))
                self_distance, required_doors = distances[pair]
                required_keys = {d.lower() for d in required_doors}

                if not required_keys.issubset(acquired_keys):
                    continue
                new_acquired_keys = set(acquired_keys)
                new_acquired_keys.add(tgt)

                rest = tgtset.difference({tgt})

                d,p = get_optimal_path(dst,rest, tgt, new_acquired_keys, distances)

                if not d or not p:
                    continue

                dd = d+self_distance
                pp = start+p
                if distance == None or dd < distance:
                    distance = dd
                    path = pp

        dst[key] = [distance,path]

    return dst[key]

def get_shortest_path_smart(mapp, pos):

    keyset = {k for k in lowercase_alpha if mapp.get_tile_locations(k)}

    locations=sorted(list(keyset)+['@'])
    distances = compute_distances(locations,mapp)
    print("distances computed")

    # for k in distances:
    #     print(k,distances[k][0], ''.join(sorted(distances[k][1])))

    optimal_paths = {}

    dist,path = get_optimal_path(optimal_paths,keyset, '@',set(), distances)
    return path, dist



if len(sys.argv) != 2:
    print("Insufficient arguments!")
    sys.exit(1)

the_input = open(sys.argv[1]).read().splitlines()

mapp = Map(the_input)

mapp.print()

initial_pos = mapp.get_tile_locations('@')[0]
print("Pos:",initial_pos)

# path,dist = get_shortest_path_bruteforce(mapp,initial_pos)

# print()
# print("Shortest path:",''.join(path))
# print("Distance:     ",dist)

path,dist = get_shortest_path_smart(mapp,initial_pos)

print()
print("Shortest path:",path)
print("Distance:     ",dist)

File: day18/day18.cpp
#include <algorithm>
#include <array>
#include <chrono>
#include <cmath>
#include <cstdio>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <thread>
#include <vector>

using coord_t = std::array<int,2>;

void get_surrounding_positions(std::vector<coord_t> &surrounding, coord_t pos, int height, int width)
{
  for(int dim=0; dim<2; dim++) {
    for( int o=-1; o<2; o+=2) {
      coord_t newpos{pos};
      newpos[dim] += o;
      if(newpos[0] >= 0 && newpos[0] < height && newpos[1] >= 0 && newpos[1] < width)
        surrounding.push_back(newpos);
    }
  }
}

using map_path_t = std::tuple<std::vector<char>,int>;

// Compares two paths according to length
bool compare_paths(const map_path_t &p1, const map_path_t &p2)
{
  return std::get<1>(p1) < std::get<1>(p2);
}

class Map
{
public:
  Map(const char* infile) {
    std::ifstream is(infile);

    std::string line;
    int nlines = 0;
    int linelen = -1;
    while (std::getline(is, line)) {
      if(linelen < 0) {
        linelen = line.size();
      }

      int oldsize = data.size();
      data.resize(oldsize+line.size());
      for(int i=0; i<line.size(); i++) {
        data[oldsize+i] = line[i];
      }
      nlines++;
    }

    this->width = linelen;
    this->height = nlines;
  }

  void print() const {
    for(int i=0; i< this->height; i++) {
      for (int j=0; j<this->width; j++) {
        putchar(this->data[i*this->width+j]);
      }
      printf("\n");
    }
  }

  char get(coord_t p) const {
    return this->data[p[0]*this->width +p[1]];
  }

  void set(coord_t p, char c) {
    this->data[p[0]*this->width +p[1]] = c;
  }

  coord_t get_first_location(char c) const {
    for(int i=0; i< this->height; i++) {
      for (int j=0; j<this->width; j++) {
        if(this->data[i*this->width+j]==c)
          return coord_t{i,j};
      }
    }

    return coord_t{-1,-1};
  }

  void get_all_locations(std::vector<coord_t>& locations, char c) const {
    for(int i=0; i< this->height; i++) {
      for (int j=0; j<this->width; j++) {
        if(this->data[i*this->width+j]==c)
          locations.push_back(coord_t{i,j});
      }
    }
  }

  void get_accessible_keys(std::vector<std::tuple<char,coord_t,int>> &accessible, coord_t pos) const {

    Map map_copy{*this};

    std::vector<coord_t> wavefronts;
    wavefronts.push_back(pos);

    int nsteps = 1;

    while(wavefronts.size() > 0) {

      std::vector<coord_t> wavefronts_new;
      for(coord_t wpos : wavefronts) {
        std::vector<coord_t> sur;
        get_surrounding_positions(sur, wpos,this->height,this->width);

        for(coord_t pt : sur) {
          char t = map_copy.get(pt);

          if(t == '#' || (t>= 'A' && t<='Z'))
            continue;

          if(t >='a' and t<='z') {
            // we found a key!
            accessible.push_back({t,pt,nsteps});
          }
          else if(t == '.') {
            map_copy.set(pt,'@');
            wavefronts_new.push_back(pt);
          }
        }
      }

      std::swap(wavefronts,wavefronts_new);
      nsteps += 1;
    }
  }


  map_path_t get_shortest_path(coord_t pos, std::vector<char> history = {}, int lvl=0) const
  {
    // printf("%*sEntering get_shortest_path (lvl %d)", 2*lvl,"", lvl);
    // for(auto c : history)
    //   printf(" %c >",c);
    // printf("\n");

    std::vector<std::tuple<char,coord_t,int>> accessible_keys;
    get_accessible_keys(accessible_keys,pos);

    std::vector<map_path_t> paths;

    for(const auto &ak : accessible_keys) {

      const auto& [key, kpos, kdist] = ak;
      Map map_copy(*this);

      // move position and remove key
      map_copy.set(pos,'.');
      map_copy.set(kpos,'@');
      // remove door/block
      coord_t door_location = map_copy.get_first_location(toupper(key));
      if(door_location[0] != -1 && door_location[1] != -1) {
        map_copy.set(door_location,'.');
      }

      std::vector<char> newhist{history};
      newhist.push_back(key);
      auto [path,dist] = map_copy.get_shortest_path(kpos, newhist, lvl+1);
      path.push_back(key);

      paths.push_back({ path, dist+kdist });
    }

    if(paths.size() == 0) {
    // printf("%*sExiting get_shortest_path  (lvl %d)\n", 2*lvl,"", lvl);
      return {{},0};
    }

    std::sort(paths.begin(),paths.end(),compare_paths);

    // printf("%*sExiting get_shortest_path  (lvl %d)\n", 2*lvl,"", lvl);
    return paths[0];
  }


private:
  std::vector<char> data;
  int width;
  int height;
};


int main(int argc, char **argv) {

  if (argc < 2) {
    printf("Too few arguments\n");
    return 1;
  }

  const char *infile = argv[1];

  Map map(infile);

  map.print();

  coord_t initial_pos = map.get_first_location('@');
  printf("Initial pos: [%d, %d]\n",initial_pos[0],initial_pos[1]);

  auto [keys,dist] = map.get_shortest_path(initial_pos);

  std::cout << std::endl;
  std::cout << "Shortest path:";
  for(auto k : keys) {
    std::cout << "  " << k;
  }
  std::cout << std::endl;
  std::cout << "Distance:     " << dist << std::endl;

  return 0;
}
